### PR TITLE
Version with crankshaft, dihedral movements (with blocks) and steepest descent minimization

### PR DESCRIPTION
**Pull request by JosePereiraUA**
TODO:
- Still lacks more/better examples.

### CODE REVIEW

File: src/Aux/misc.jl [hunk 1]
```diff
@@ -11,12 +11,45 @@ julia> Aux.rotation_matrix_from_axis_angle([1.1, 2.2, 3.3], π/2)
...
+Calculates the dihedral angle produced between a1, a2, a3 and a4, in degrees, by default. If `radians` flag is set to true, the output is in radians.
+
+# Examples
+```julia-repl
+julia> Aux.calc_dih_angle([1.0, 1.0, 1.0], [2.1, 2.1, 2.1], [3.0, 2.0, 5.0], [5.0, 5.0, 5.0], radians = true)
+3.141592653589793
+```
+See also: [`apply_initial_conf!`](@ref Common)
+"""
+function calc_dih_angle(a1::Vector{Float64}, a2::Vector{Float64}, a3::Vector{Float64}, a4::Vector{Float64};
```
Comment (sergiomsantos) [inline src/Aux/misc.jl]:
```suggestion
function calculate_dihedral(a1::Vector{Float64}, a2::Vector{Float64}, a3::Vector{Float64}, a4::Vector{Float64})
```
the function should always return radians

File: src/Aux/misc.jl [hunk 2]
```diff
@@ -11,12 +11,45 @@ julia> Aux.rotation_matrix_from_axis_angle([1.1, 2.2, 3.3], π/2)
...
+    n2 = cross(b2, b3)
+    x = dot(cross(n1, n2), b2) / sqrt(dot(b2, b2))
+    y = dot(n1, n2)
+    v = atan(x, y)
+
+    if radians
+        return v
+    else
+        return rad2deg(v)
+    end
```
Comment (sergiomsantos) [inline src/Aux/misc.jl]: delete this entire `if block`

File: src/Common/callback.jl [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+julia> Common.CallbackObject(Print.as_xyz)
+CallbackObject(callback=Print.as_xyz, freq=1)
+```
+See also: [`Print.as_xyz`](@ref Print) [`@cbcall`](@ref)
+"""
+mutable struct CallbackObject
+    callback::Function
+    freq::Int64
+end
+CallbackObject(callback::Function; freq::Int64 = 1) = CallbackObject(callback, freq)
```
Comment (sergiomsantos) [inline src/Common/callback.jl]:
```suggestion
CallbackObject(callback::Function) = CallbackObject(1, callback)
```
se apenas definirmos a função, a frequência é sempre 1

File: src/Common/callback.jl [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+julia> Common.CallbackObject(100, Print.as_xyz)
+CallbackObject(callback=Print.as_xyz, freq=100)
+
+julia> Common.CallbackObject(Print.as_xyz)
+CallbackObject(callback=Print.as_xyz, freq=1)
+```
+See also: [`Print.as_xyz`](@ref Print) [`@cbcall`](@ref)
+"""
+mutable struct CallbackObject
+    callback::Function
```
Comment (sergiomsantos) [inline src/Common/callback.jl]: inverte a ordem:
- primeiro freq
- depois callback

File: src/Common/macros.jl [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+
+(Macro) Create a [`CallbackObject`](@ref) with the given function `f` and output frequency `freq`. 
+
+# Examples
+```julia-repl
+julia> @Common.callback my_callback1 10
+CallbackObject(callback=my_callback1, freq=10)
+```
+"""
+macro callback(f, freq)
```
Comment (sergiomsantos) [inline src/Common/macros.jl]:
```suggestion
macro callback(args...)
    ex = :(Common.CallbackObject($(args...)))
    esc(ex)
end
```
assim podemos definir o `CallbackObject` com número variável de argumentos, sendo que se apenas passarmos a função, então `freq=1`

**Review by sergiomsantos — DISMISSED** [review]

**Comment by JosePereiraUA** [comment]
The Steepest Descent example still needs work. Either a secondary PDB file is read and its coordinates applied to the "normal" PDB file (with the molecule straight), or a new JSON file needs to be created (because the current one has the atom in the old ProtoSyn order). Tried to use the old ProtoSyn to apply the native 1CTF coordinates, but some errors were occurring.

Bom fim de semana!

**Pull request merged**